lib/AST/ASTScopeCreation.cpp

```diff
@@ -617,7 +617,7 @@ class ScopeCreator final {
 #endif
     ASTScopeAssert(startOrder * endOrder != -1,
                    "Start order contradicts end order");
-    return startOrder + endOrder < 1;
+    return startOrder < 0 && endOrder < 0;
```

Comment:
This is a semantics change, though. Exact matching ranges, for example, should be considered "not after".

(I'm also a little confused by this whole idea. One range nested entirely inside another will trip the assert.)

Reply:
Yeah, it was very confusing.  I *think* that `<=` should maintain the strict weak ordering, so that seems like a better approach.  I'll change that.

Reply:
Thank you, @compnerd for working on fixing my oversight. Thank you, @jrose-apple for flagging me. I believe that this change is incorrect, but so is the existing code I wrote. 

The ordering here is used for sorting, which is used in `ASTScopeImpl::findChildContaining`. That function compares a location against the `End` of the ranges only. What is needed is a factoring: a function that does the comparison and is used by *both* the sort routing and the `findChildContaining` routine. Otherwise, we'll have hiding places for bugs. The factored routine should just use the `End` locations. (I originally chose that because @DougGregor 's prototype used that.)

In addition to factoring (and keeping the assertion), I wonder if the assertion should be turned into a test that crashes the compiler, even in production runs. The rationale would be that a miscompile will result from incorrect lookups when the ranges overlap. What do you both think?

Reply:
I'm not really certain how to refactor this.  The `findChildContaining` simply queries the `SourceManager`, and the comparator doesn't really generalise very much.  Did you have something in mind?  I'd like to get this resolved ASAP, this is breaking the Windows build.